matchms/SpectrumProcessor.py

```diff
@@ -91,32 +92,41 @@ def add_custom_filter(self, filter_function, filter_params=None):
             filter_func.__name__ = filter_function.__name__
             self.filters.append(filter_func)
 
-    def process_spectrum(self, spectrum):
+    def process_spectrum(self, spectrum,
+                         processing_report: Optional["ProcessingReport"] = None):
         """
         Process the given spectrum with all filters in the processing pipeline.
 
         Parameters
         ----------
         spectrum : Spectrum
             The spectrum to process.
-        
+        processing_report:
+            A ProcessingReport object When passed the progress will be added to the object.
+
         Returns
         -------
         Spectrum
             The processed spectrum.
         """
         if self.filters == []:
             raise TypeError("No filters to process")
+        if processing_report is not None:
+            processing_report.counter_number_processed += 1
         for filter_func in self.filters:
-            spectrum = filter_func(spectrum)
-            if spectrum is None:
+            spectrum_out = filter_func(spectrum.clone())
```

Comment:
Here I added .clone() By adding it here it will not be necessary anymore to do this in each filter function. Which I think would be a nice and saver way. However, with the current filter functions this will result in cloning twice, which is time consuming. We can remove this here, remove the spectrum.clone() from the filter functions, or accept the extra runtime for extra safety. What do you think?

Reply:
Even though it looks like a massive code duplication, I would leave the clone() part in each filter because that is what I would expect an individual filter function to do (and not to change the spectrum object itself). 
There is still the option to streamline this more in the near future as you suggested Niek. I also just added some ideas to your issue #434 on this.

Reply:
Here, we could hence skip it (to avoid cloning twice).